labman/db/process.py

```diff
@@ -1751,22 +1773,44 @@ def create(cls, user, plate, concentrations):
                         (quantitated_composition_id, upstream_process_id,
                          raw_concentration)
                      VALUES (%s, %s, %s)"""
-            sql_args = []
-            layout = plate.layout
 
-            for p_row, c_row in zip(layout, concentrations):
-                for well, conc in zip(p_row, c_row):
-                    if well is not None:
-                        sql_args.append([well.composition.composition_id,
-                                         instance.id, conc])
+            if plate is not None:
```

Comment:
So this is basically uniting the plate (per composition) and pool (single composition of a pool) quantifications? And it infers what it's doing based on the presence of a value for `plate`?

Reply:
@tanaes Yes, precisely.  The new _create method unites the previous contents of create and create_manual (the create and create_manual methods still exist, so as not to alter the public interface for the class, but they each just call _create under the hood with the appropriate set of inputs for their functionality).  The reason for uniting these two methods is that they duplicated most of their functionality (including several string-literal SQL queries, which is a maintenance and bug-finding nightmare!) and I would have had to modify each of them separately but in exactly the same way in order to allow storage of notes, which is a design red flag for me.

As you noted,  _create decide whether the quantification is per-plate or per-pool based on whether the plate argument is provided.  This is perhaps not ideal (because, for example, there's no validation happening to ensure that of the format of the quantifications argument matches what it should be given the presence or absence of the plate argument); however,  keep in mind that _create is *only* called from within create or create_manual.  Those two public methods are therefore responsible for making it clear to the user what the acceptable combination of inputs is for each of the two ways of doing a quantification, and _create is assuming it is being passed sensible inputs from one of those two gate-keepers.

Reply:
Awesome, thanks for finding that and for the clarification!